examples/device.rs

```diff
@@ -24,23 +24,17 @@
 #![no_main]
 #![no_std]
 
-extern crate cortex_m;
-#[macro_use]
-extern crate cortex_m_rt as rt;
-extern crate cortex_m_semihosting as sh;
-#[macro_use]
-extern crate stm32f103xx;
-extern crate panic_semihosting;
+#[allow(unused_extern_crates)]
+extern crate panic_halt;
 
 use core::fmt::Write;
 
 use cortex_m::peripheral::syst::SystClkSource;
-use rt::ExceptionFrame;
-use sh::hio::{self, HStdout};
-use stm32f103xx::Interrupt;
-
-entry!(main);
+use cortex_m_rt::entry;
+use cortex_m_semihosting::hio::{self, HStdout};
+use stm32f30x::{interrupt, Interrupt};
```

Comment:
I would prefer if we could move the turn the device specific `use` lines into generic ones and define the crates to be used in the end in src/lib.rs, aliasing the device specific peripheral registers and HAL to `device` and `hal` respectively so (in Rust 2018) those would then read
```
use crate::device::<whatever>;
use crate::hal::<whatever>;
```
instead in the examples. This makes it a lot easier to try them on different devices.

Reply:
I don’t think example programs should depend on src/lib.rs of this crate.

Users typically use examples in their own crates.  In that case, they can’t touch src/lib.rs of cortex-m-rt.

In other words, I expect I can do the following thing:
1) copy one of the examples into their own main.rs.
2) edit cargo.toml .
3) build successfully

Adding one indirection of ‘use’ clauses will do harm in this use case.

Reply:
I don't quite see your point here I'm afraid. That's exactly what I would hope to enable by generally using universal names instead of calling device specific crates because those are the ones that one always has to change in each source file. If you want to use an example in your own crate you just have to alias the correct crates in your src/lib.rs and if the hardware is similar enough it will just run.